plugins/cwe_checker_emulation/cwe_checker_emulation.ml

```diff
@@ -0,0 +1,178 @@
+open Core_kernel
+open Bap.Std
+open Bap_primus.Std
+open Bap_future.Std
+open Graphlib.Std
+open Monads.Std
+open Format
+open Ppx_jane
+open Cwe_checker_core
+
+include Self()
+
+let pp_id = Monad.State.Multi.Id.pp
+
+module Machine = struct
+  type 'a m = 'a
+  include Primus.Machine.Make(Monad.Ident)
+end
+open Machine.Syntax
+
+module Main = Primus.Machine.Main(Machine)
+module Interpreter = Primus.Interpreter.Make(Machine)
+module Linker = Primus.Linker.Make(Machine)
+module Env = Primus.Env.Make(Machine)
+module Lisp = Primus.Lisp.Make(Machine)
+module Eval = Primus.Interpreter.Make(Machine)
+
+(** this array collects the observed primus events*)
+let collected_events = ref ([||])
+
+(** Converts a hexadecimal string representation of
+an address to an integer. *)
+let convert_location loc =
+  match (Str.split (Str.regexp ":") loc) with
+  | fst::snd::[] -> Int.of_string ("0x" ^ snd)
+  | _ -> failwith "Could not parse location"
+
+(** Converts a list of hexadecimal strings to a
+list of integers. *)
+let convert_location_list loc_list =
+  let locs = ref [] in
+  Sexplib__Sexp_with_layout.List.iter loc_list ~f:(fun x -> locs := (convert_location @@ Sexp.to_string x)::(!locs));
+  !locs
+
+(** Analyze events and report to the user. *)
+let analyze_events _ =
+  let location_tbl = Hashtbl.create (module String) in
+  Array.iter ~f:(fun (p, ev) ->
+      begin
+        match ev with
+        |  Sexp.Atom _ -> failwith "Sexp.Atom not expected in report_events."
+        |  Sexp.List [Sexp.Atom location_id; Sexp.List location_list] -> Hashtbl.add_exn location_tbl location_id (convert_location_list location_list)
+        |  Sexp.List incident -> Incident_reporter.report incident location_tbl
+      end) !collected_events
+
+(** Just adds the observed Primus events to the collected_events array. *)
+let collect_events p ev =
+  collected_events := Array.append !collected_events [|(p, ev)|]
+
+(* Most functions beyond here have been taken and adjusted from BAP's Primus plugins*)
+
+let string_of_name = function
+  | `symbol s -> s
+  | `tid t -> Tid.to_string t
+| `addr x -> Addr.string_of_value x
+
+(** Collects all entry points of the program. *)
+let entry_point_collector = object
+  inherit [tid list] Term.visitor
+  method! enter_term _ t entries =
+    if Term.has_attr t Sub.entry_point
+    then Term.tid t :: entries
+    else entries
+  end
+
+(** Wrapper function around entry_point_collector. *)
+let entry_points prog =
```

Comment:
There is a similar function in symbol_utils, except it also adds the main function as entry point. I assume adding the main function here would break something? This function should probably be moved to symbol_utils.

Reply:
I am using a function from Symbol_utils now.